model/riscv_insts_zicsr.sail

```diff
@@ -31,7 +31,11 @@ function clause execute CSR(csr, rs1, rd, is_imm, op) = {
   else if not(ext_check_CSR(csr, cur_privilege, isWrite))
   then { ext_check_CSR_fail(); RETIRE_FAIL }
   else {
-    let csr_val = read_CSR(csr); /* could have side-effects, so technically shouldn't perform for CSRW[I] with rd == 0 */
+    let csr_val : xlenbits = match (op, rd) {
+      /* CSRRW should not generate read side-effects if rd == 0 */
+      (CSRRW, 0b00000) => zeros(),
+      _                => read_CSR(csr),
+    };
```

Comment:
We did it like this, which I think is a bit nicer:

```
  // CSRRW[I] always writes. CSRRS[I] and CSRRC[I] write if the
  // source register (or immediate) is not x0 (or 0).
  let isWrite = op == CSRRW | rs1 != zeros();
  // CSRRS[I] and CSRRC[I] always read. CSRRW[I] reads if the destination
  // register is not x0.
  let isRead = op != CSRRW | rd != zeros();
...
    let csr_val = if isRead then read_CSR(csr) else zeros();
```

(Sorry; on my list of things to upstream.)

Reply:
That works too.  I would have preferred 'isCSRWrite' and 'isCSRRead' (or 'is_CSR_Write' ...) to be less confusing, since rs1 and rd are also being read and written.  I can edit the commit to the above if you agree.

Reply:
Yeah that makes sense.